tests/TurboMqtt.Tests/End2End/TcpMqtt311End2EndSpecs.cs

```diff
@@ -46,8 +48,156 @@ protected override void AfterAll()
+            Func<(IMemoryOwner<byte> buffer, int estimatedSize), bool> pushMessage, 
+            Func<Task> closingAction,
+            ILoggingAdapter log,
+            TimeSpan? heartbeatDelay = null) 
+            : base(pushMessage, closingAction, log, heartbeatDelay)
+        {
+            _onConnectCallback = onConnectCallback;
+        }
+
+        public override void HandlePacket(MqttPacket packet)
+        {
+            if (packet.PacketType == MqttPacketType.Connect)
+            {
+                var connect = (ConnectPacket)packet;
+                ClientIdAssigned.TrySetResult(connect.ClientId);
+                _onConnectCallback(connect);
+                return;
+            }
+            base.HandlePacket(packet);
+        }
+    }
+    
+    private sealed class ConfigurableFakeServerFactory: IFakeServerHandleFactory
+    {
+        private readonly Func<Func<(IMemoryOwner<byte> buffer, int estimatedSize), bool>, Func<Task>, ILoggingAdapter,
+            TimeSpan?, IFakeServerHandle> _onCreateHandlerCallback;
+
+        public ConfigurableFakeServerFactory(
+            Func<Func<(IMemoryOwner<byte> buffer, int estimatedSize), bool>, Func<Task>, ILoggingAdapter,
+                TimeSpan?, IFakeServerHandle> onCreateHandlerCallback)
+        {
+            _onCreateHandlerCallback = onCreateHandlerCallback;
+        }
+
+        public IFakeServerHandle CreateServerHandle(
+            Func<(IMemoryOwner<byte> buffer, int estimatedSize), bool> pushMessage, 
+            Func<Task> closingAction,
+            ILoggingAdapter log,
+            MqttProtocolVersion protocolVersion = MqttProtocolVersion.V3_1_1, 
+            TimeSpan? heartbeatDelay = null)
+        {
+            return _onCreateHandlerCallback(pushMessage, closingAction, log, heartbeatDelay);
+        }
+    }
+    
+    /// <summary>
+    /// This is an edge case when the client tries to reconnect and the socket disconnected right after the CONNECT
+    /// packet was sent by the client but before the CONNACK packet were received by the client.
+    ///
+    /// The bug was that the ClientAcksActor were stuck waiting for the previous connection state to complete, blocking
+    /// immediate client reconnect attempt.
+    ///
+    /// There should be 3 connection attempts in this test, with these steps happening in sequence:
+    /// 
+    /// 1. Client connects normally to the broker
+    /// 2. Client connected successfully to the broker
+    /// 3. Socket connection lost (forcefully)
+    /// 4. Client tries to reconnect to the broker
+    /// 5. Client socket connected to the broker and sends a CONNECT packet, ClientAcksActor _pendingConnect field is set
+    /// 6. Socket connection lost (forcefully)
+    /// 7. ClientAcksActor _pendingConnect field is reset by a Reconnect message
+    /// 8. Client tries to reconnect to the broker
+    /// 9. Client connected successfully to the broker
+    /// </summary>
```

Comment:
Regression test description, what the bug was and what supposed to happen after the fix.

Reply:
I think technically you can skip steps 1-3 - this failure can happen the first time a client connects too.